src/gen/async_generate_data_gemini.py

```diff
@@ -0,0 +1,99 @@
+
+
+# prompt template can be loaded from a `.toml` file.
+def craft_prompt(instruction, response):
+    prompt = """
+Generate a series of (instruction, response) pairs that are similar in context and structure to the example provided below. Each pair should consist of a concise instruction followed by an appropriate, detailed response. The instruction should pose a clear task or question, while the response should provide a comprehensive answer or solution that could be understood by someone with a basic understanding of the subject.
+
+Example pair:
+
+Instruction: {instruction}
+Response: {response}
+
+Your task is to generate more pairs that maintain this level of clarity and detail. The topic is Coding. Ensure that the responses are informative and accurate, suitable for an educational context.
+
+Store the generated pairs in JSON format, with each pair as an object within an array. Each object should have two key-value pairs: "instruction" and "response". For instance:
+
+[{{"instruction": "text", "response": "text"}}, {{"instruction": "text", "response": "text"}}, ...]
+
+Remember to maintain consistency in the format and ensure the generated pairs are diverse and cover a broad range of subjects. You must return the response
+in the asked format and you must not add any additional text in your response.
+"""
+    return prompt.format(instruction=instruction, response=response)
+
+
+async def generate_text(prompt):
+    url = (
+        url
+    ) = f"https://generativelanguage.googleapis.com/v1beta/models/{MODEL_NAME}:generateContent?key={GOOGLE_API_KEY}"
+    data = {"contents": [{"parts": [{"text": prompt}]}]}
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(url, json=data) as response:
+            if response.status == 200:
+                result = await response.json()
+                return result
+            else:
+                print(f"Error: {response.status}")
+                return None
+
+
+async def main():
+    # craft prompts
+    total_original_samples = len(EXISTING_DATASET)
+    random_indices = np.random.randint(0, total_original_samples, size=(NUM_DATASET_SAMPLES))
+    prompts = []
+    print("Preparing candidate prompts.")
+    for random_index in random_indices:
+        sample = EXISTING_DATASET[int(random_index)]["messages"]
+        instruction = sample[0]["content"]
+        response = sample[1]["content"]
+        prompt_for_sample = craft_prompt(instruction, response)
+        prompts.append(prompt_for_sample)
+
+    # create tasks
+    tasks = []
+    print("Creating tasks for asyncio.")
+    for prompt in prompts:
+        tasks.append(generate_text(prompt))
+        # rate-limiting
+        await asyncio.sleep(REQUEST_INTERVAL)
+
+    # fire execution
+    results = await asyncio.gather(*tasks)
+
+    # collate results
+    print("All tasks done. Collating results.")
+    for i, result in enumerate(results):
```

Comment:
curious why dumping to external JSON files instead of just pushing them to HF Dataset repository?

Reply:
Please check the notebook. It's all commented on there.

Reply:
> This is better run from a script rather than a notebook. The following script can also be refactored later. I also propose to first collect all the Gemini responses and serialize them in a simple JSON file. Once that's done it will be pretty easy to use another script to format the results and have them compatible with datasets.

Cool. This could be a decision choice. 
Like you said this project is still in its infancy, things could be changed. Once we implement all pieces, let's play with them, and make some modifications in the UX/DX perspective. Does it sounds OK?

Reply:
Yes 💯 percent.